redis/src/aio/connection.rs

```
@@ -508,3 +510,101 @@ pub(crate) async fn connect_simple<T: RedisRuntime>(
+/// * `F` - Closure type taking (C, Pipeline) and returning Fut
+/// * `Fut` - Future returning Result<Option<T>, RedisError>
+///
+/// # Examples
+///
+/// ```rust,no_run
+/// use redis::{AsyncCommands, RedisResult, pipe};
+///
+/// async fn increment(con: redis::aio::ConnectionManager) -> RedisResult<isize> {
+///     let key = "my_counter";
+///     redis::aio::transaction_async(con, &[key], |mut con, mut pipe| async move {
+///         //  Get the current value
+///         let val: isize = con.get(key).await?;
+///         //   Create pipeline that increments and returns new value
+///         pipe.set(key, val + 1)
+///             .ignore()
+///             .get(key)
+///             .query_async(&mut con)
+///             .await
+///     })
+///     .await
+/// }
+/// ```
+///
+/// # Notes
+///
+/// - The closure provided to this function may be executed multiple times if the
+///   watched keys are modified by other clients between the WATCH and EXEC commands
+/// - Side effects in the closure should be idempotent since they may occur multiple times
+/// - The transaction is automatically rolled back if the closure returns an error
+/// - Returns the final result once the transaction succeeds
+///
+/// For more details on Redis transactions, see the [Redis documentation](https://redis.io/topics/transactions)
+pub async fn transaction_async<
+    Connection: ConnectionLike + Clone,
+    K: ToRedisArgs,
+    T: FromRedisValue,
+>(
+    mut connection: Connection,
+    keys: &[K],
+    pipeline: &mut Pipeline,
+) -> Result<T, RedisError> {
+    let mut transaction_pipeline = pipe(); // a long lived value that will be used to build the transaction pipeline
+    transaction_pipeline.cmd("WATCH").arg(keys).cmd("MULTI");
+
+    std::mem::take(&mut pipeline.commands) // consume the pipeline and give into_iter() ownership of the commands attribute
+        .into_iter()
+        .filter(|cmd| cmd.data != b"UNWATCH")
```

I prefer to return an error here, instead of silently ignoring this